### DISCUSSION TITLE
Delay client React render

### DISCUSSION THREAD

**Discussion by Oikio** [vercel/next.js#14879] [Help] [answered]
I have asked same question on https://github.com/vercel/next.js/issues/6817#issuecomment-646560793, which is closed and of course didn't get any attention, maybe community might help?

Here's a copy paste with some additional information:

There is an issue with runtime styled-components theming.
Themes are not serializable, so I can't use `getInitialProps`, but I don't want to require them all in bundle, they will bloat it.

So I decided to go with conditional require on server side and with dynamic import on client, but as one would expect got blinking due to first render not having a theme yet.
Works as follows:  renders properly -> React hydrates synchronously -> renders without theme (view is null in this case, can't render without theme) -> renders with theme.

On any other app, I would just load everything I need before initializing React, how can I do it with next then?

**Comment by oleg-preply** [comment]
Found working, but not really nice, workaround:

The only thing Next's `await`s for before hydrating app is [`await`ing for loading dynamic chunks](https://github.com/vercel/next.js/blob/a0c683208abdce84934061f6efd2b90de52d99b9/packages/next/client/index.js#L261) (only those that were SSR-ed).
So I get into that moment and ask to wait also for my custom Promise:

```javascript
// inlined somewhere on the top of html doc
const myCustomPromise = new Promise(...);

Object.defineProperty(window, '__NEXT_PRELOADREADY', {
        set(value) {
            this.__NEXT_PRELOADREADY_ORIGINAl = value;
        },
        get() {
            return (dynamicIds: string[]) =>
                this.__NEXT_PRELOADREADY_ORIGINAl(dynamicIds).then(() => myCustomPromise);
        },
    });
```

But of course, using Next's internals is not really good idea... But smth working, at least now :)

**Reply by Oikio** [reply]
That's something! At least it can be used in my case. I guess we need feature request for this one, maybe it won't be closed this time.

**Reply by timneutkens** [reply]
I'd highly recommend not doing this as you're basically making your app slow, this is an internal property used for code loading in a very specific case and should not be used by users. It's extremely likely to break your app down the line with no way to fix your application. 

Unmarked the answer.

**Reply by timneutkens** [reply]
https://github.com/vercel/next.js/discussions/14879#discussioncomment-46443

**Reply by oleg-preply** [reply]
@timneutkens that's clear and generally totally agree :ok_hand: 

I'm executing that script only for users, who don't have some native APIs implemented in their browsers (like `Intl.PluralRules`) - so the rest of them are not affected by waiting for downloading that additional polyfill.

In my specific case benefits superseded risks...
> It's extremely likely to break your app down the line with no way to fix your application.

in that case I will need to load all the polyfills unconditionally for all the users :( - it also makes app slower

**Reply by chrisrichard** [reply]
@oleg-preply we have the same case that we'd like to use it for. We're seeing low-frequency JS errors around random Intl stuff missing in older Safari versions. We don't want to bundle polyfills for new browsers and it's better to slow down Safari a bit if the alternative is the app doesn't load.

**Reply by eemeli** [reply]
@oleg-preply Regarding `Intl.PluralRules`, if you know the subset of locales that you need to support, you should be able to use `intl-pluralrules/factory` to build a polyfill of minimal size: https://github.com/eemeli/intl-pluralrules#factory

I hope this is something that you're already doing?

**Comment by jnv** [comment]
@Oikio You may want to also check out the plugins RFC #9133 which covers lifecycle methods, especially this one:

```js
// on-init-client.js
export default async function onInitClient({ router }): void {
  // called on client init, before hydration
}
```

It is not an immediate solution but I think this use case could be eventually covered.

**Reply by Oikio** [reply]
That's looks like something I could use if it was implemented =)

**Comment by timneutkens** [comment, marked as answer]
> but I don't want to require them all in bundle, they will bloat it.

You'll likely want to use `next/dynamic` https://nextjs.org/docs/advanced-features/dynamic-import#basic-usage for a context provider that can then be switched out between pages for example.

**Reply by Oikio** [reply]
Dynamic loading was my first approach:

> So I decided to go with conditional require on server side and with dynamic import on client, but as one would expect got blinking due to first render not having a theme yet.

**Reply by timneutkens** [reply]
It's likely that you implemented it wrong, did you use `next/dynamic` as per the documentation?

**Reply by Oikio** [reply]
No, I have a styled-components theme, not a component. But I see what you mean, not sure though how to express it in code with `dynamic()` syntax.

**Reply by timneutkens** [reply]
Something like this should work (haven't tried it but based on the ThemeProvider docs):

```js
// components/theme1-provider.js
import {ThemeProvider} from 'styled-components'
import theme1 from '../themes/theme1.js'

export default function MyTheme1Provider({children}) {
  return <ThemeProvider theme={theme1}>{children}</ThemeProvider>
}
```

```js
// components/theme2-provider.js
import {ThemeProvider} from 'styled-components'
import theme1 from '../themes/theme2.js'

export default function MyTheme2Provider({children}) {
  return <ThemeProvider theme={theme2}>{children}</ThemeProvider>
}
```

```js
// pages/index.js
import dynamic from 'next/dynamic'

// Next.js only imports the code when the component has been rendered so this list is fine
const themeProviders = {
 theme1: dynamic(() => import('../components/theme1-provider')),
 theme2: dynamic(() => import('../components/theme2-provider'))
}

export default function HomePage() {
  // Calculate the theme name to look it up in the themeProviders list
  const themeName = 'theme1'
  const ThemeProvider = themeProviders[themeName]
  return <ThemeProvider>
    <h1>Hello World</h1>
  </ThemeProvider>
}
```

**Reply by Oikio** [reply]
@timneutkens  thank you for your help, it worked! I've added all other configurations to providers, had to refactor code though.

It's still doesn't appeal to me as much as a simple possibility to have something like `export const onInitClient = ....` in _app.tsx or something, I had to refactor all services outside of react cycle to be initialized inside react, because not only I have different themes, but also different configs for brands.

**Comment by nathggns** [comment]
There's a use case not covered by `dynamic()`, which is where you have some async code which needs to run before render on both the server and the client, and you want to still support SSR. Using dynamic will mean the result rendered by SSR won't include the path blocked by this async code. It seems in this case there is a valid need to simple delay hydration, which is easy enough when you manage hydration yourself.
